publisher-sdk/src/main/java/com/criteo/publisher/logging/LogMessage.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ *    Copyright 2020 Criteo
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.criteo.publisher.logging
+
+import android.util.Log
+
+internal data class LogMessage(
+    val level: Int = Log.INFO,
+    private val message: () -> String?,
+    val throwable: Throwable? = null
+) {
+  val formattedMessage: String? by lazy {
```

Comment:
What's the benefit of passing a lambda and using a lazy property to access the formatted message? Why don't we just let the caller pass the formatted message directly? 

Specially since we are recreating a `LogMessage` every time we call `Logger.log`, so we will never take advantage of it. Also, even if it weren't the case, is formatting a message that expensive?

More info on the real cost of lazy: https://youtu.be/dyprnRjQRM8?t=1003

Reply:
That's a cool talk.

You're right, this is premature optimization without benchmarking. It even adds legacy code with overridden equal/hashCode/toString.

We can go on a lazy solution if we detect some slowness. But we're not flooding the logcat and the biggest logs are done before and after a network call.